File: python-ai-service/models/plaid_item.py
from datetime import datetime
from typing import Optional
from beanie import Document, PydanticObjectId


class PlaidItem(Document):
    owner: PydanticObjectId
    access_token: str  # Fernet-encrypted before storage
    item_id: str
    institution_name: str
    institution_id: str
    cursor: Optional[str] = None  # Plaid transaction sync cursor
    products: list[str] = []     # products this item was linked with

    class Settings:
        name = "plaid_items"
        indexes = [
            [("item_id", 1)],         # unique index declared below
            [("owner", 1)],
        ]

File: src/lib/server/models/PlaidItem.ts
import mongoose, { Schema, model, type Document } from 'mongoose';

export interface IPlaidItem extends Document {
	owner: mongoose.Types.ObjectId;
	access_token: string; // Fernet-encrypted
	item_id: string;
	institution_name: string;
	institution_id: string;
	cursor?: string | null;
	products: string[];
}

const PlaidItemSchema = new Schema<IPlaidItem>(
	{
		owner: { type: Schema.Types.ObjectId, ref: 'User', required: true, index: true },
		access_token: { type: String, required: true },
		item_id: { type: String, required: true, index: true },
		institution_name: { type: String, required: true },
		institution_id: { type: String, required: true },
		cursor: { type: String, default: null },
		products: { type: [String], default: [] }
	},
	{ collection: 'plaid_items', timestamps: true }
);

export const PlaidItem =
	(mongoose.models.PlaidItem as mongoose.Model<IPlaidItem>) ??
	model<IPlaidItem>('PlaidItem', PlaidItemSchema);
